src/main/java/liqp/tags/For.java

```diff
@@ -83,188 +103,239 @@ private Object renderArray(String id, TemplateContext context, LNode... tokens)
+        }
+    }
 
-            context.put(id, array[i]);
 
-            forLoopMap.put(LENGTH, length);
-            forLoopMap.put(INDEX, n + 1);
-            forLoopMap.put(INDEX0, n);
-            forLoopMap.put(RINDEX, length - n);
-            forLoopMap.put(RINDEX0, length - n - 1);
-            forLoopMap.put(FIRST, first);
-            forLoopMap.put(LAST, last);
+    private boolean renderForLoopBody(TemplateContext context, StringBuilder builder, List<LNode> children) {
+        boolean isBreak = false;
 
-            List<LNode> children = ((BlockNode)block).getChildren();
-            boolean isBreak = false;
+        for (LNode node : children) {
 
-            for (LNode node : children) {
+            Object value = node.render(context);
 
-                Object value = node.render(context);
+            if(value == null) {
+                continue;
+            }
 
-                if(value == LValue.CONTINUE) {
-                    // break from this inner loop: equals continue outer loop!
-                    break;
-                }
+            if(value == LValue.CONTINUE) {
+                // break from this inner loop: equals continue outer loop!
+                break;
+            }
 
-                if(value == LValue.BREAK) {
-                    // break from inner loop
-                    isBreak = true;
-                    break;
-                }
+            if(value == LValue.BREAK) {
+                // break from inner loop
+                isBreak = true;
+                break;
+            }
 
-                if (value != null && value.getClass().isArray()) {
+            if(super.isArray(value)) {
 
-                    Object[] arr = (Object[]) value;
+                Object[] arr = super.asArray(value);
 
-                    for (Object obj : arr) {
-                        builder.append(String.valueOf(obj));
-                    }
-                }
-                else {
-                    builder.append(super.asString(value));
+                for (Object obj : arr) {
+                    builder.append(String.valueOf(obj));
                 }
-            }
-
-            if(isBreak) {
-                // break from outer loop
-                break;
+            } else {
+                builder.append(super.asString(value));
             }
         }
-
-        context.put(CONTINUE, continueIndex + 1, true);
-
-        return builder.toString();
+        return isBreak;
     }
 
-    private Object renderRange(String id, TemplateContext context, LNode... tokens) {
+    private Object renderRange(String id, TemplateContext context, String tagName, boolean reversed, LNode... tokens) {
 
         StringBuilder builder = new StringBuilder();
 
-        // attributes start from index 5
-        Map<String, Integer> attributes = getAttributes(5, context, tokens);
+        // attributes start from index 6
+        Map<String, Integer> attributes = getAttributes(7, context, tagName, tokens);
```

Comment:
The comment `attributes start from index 6` does not match the code: `getAttributes(7, ...)`

Reply:
Fixed.